### PR TITLE
feat: get transactions from mem pool ordered by fee

### PR DESCRIPTION
**Pull request by vasild**
## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (improve a current implementation without adding a new feature or fixing a bug)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Build (changes that affect the build system)
- [ ] Docs (documentation only changes)
- [ ] Test (adding missing tests or fixing existing tests)
- [ ] Other... Please describe:

## Checklist

- [x] I have read the [CONTRIBUTING](https://docs.ark.io/developers/guidelines/contributing.html) documentation
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by vasild** [comment]
Beware:

before this patch:
```
      ✓ multiple additions and retrievals (646ms)
      ✓ delete + add after sync (173ms)
```

after this patch:
```
      ✓ multiple additions and retrievals (5779ms)
      ✓ delete + add after sync (1306ms)
```

**Comment by vasild** [comment]
This implements the first part of https://github.com/ArkEcosystem/core/issues/1061

File: packages/core-transaction-pool-mem/lib/mem-pool-transaction.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    return this.transaction_
+  }
+
+  get sequence () {
+    assert.notEqual(this.sequence_, undefined)
+    return this.sequence_
+  }
+
+  set sequence (seq) {
+    assert.strictEqual(this.sequence_, undefined)
```
Comment (faustbrian) [inline packages/core-transaction-pool-mem/lib/mem-pool-transaction.js]:
```suggestion
    assert.strictEqual(this._sequence, undefined)
```

File: packages/core-transaction-pool-mem/lib/mem-pool-transaction.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    }
+  }
+
+  get transaction () {
+    return this.transaction_
+  }
+
+  get sequence () {
+    assert.notEqual(this.sequence_, undefined)
+    return this.sequence_
```
Comment (faustbrian) [inline packages/core-transaction-pool-mem/lib/mem-pool-transaction.js]:
```suggestion
    return this._sequence
```

File: packages/core-transaction-pool-mem/lib/mem-pool-transaction.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    this.transaction_ = transaction
+
+    if (sequence !== undefined) {
+      assert(sequence instanceof Number)
+      this.sequence_ = sequence
+    }
+  }
+
+  get transaction () {
+    return this.transaction_
```
Comment (faustbrian) [inline packages/core-transaction-pool-mem/lib/mem-pool-transaction.js]:
```suggestion
    return this._transaction
```

File: packages/core-transaction-pool-mem/lib/mem-pool-transaction.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * A normal transaction + a sequence number used to order by insertion time.
+ */
+class MemPoolTransaction {
+  constructor (transaction, sequence) {
+    assert(transaction instanceof Transaction)
+    this.transaction_ = transaction
+
+    if (sequence !== undefined) {
+      assert(sequence instanceof Number)
+      this.sequence_ = sequence
```
Comment (faustbrian) [inline packages/core-transaction-pool-mem/lib/mem-pool-transaction.js]:
```suggestion
      this._sequence = sequence
```

File: packages/core-transaction-pool-mem/lib/mem-pool-transaction.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+const { Transaction } = require('@arkecosystem/crypto').models
+
+/**
+ * A mem pool transaction.
+ * A normal transaction + a sequence number used to order by insertion time.
+ */
+class MemPoolTransaction {
+  constructor (transaction, sequence) {
+    assert(transaction instanceof Transaction)
+    this.transaction_ = transaction
```
Comment (faustbrian) [inline packages/core-transaction-pool-mem/lib/mem-pool-transaction.js]:
```suggestion
    this._transaction = transaction
```

File: packages/core-transaction-pool-mem/lib/mem-pool-transaction.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  }
+
+  get sequence () {
+    assert.notEqual(this.sequence_, undefined)
+    return this.sequence_
+  }
+
+  set sequence (seq) {
+    assert.strictEqual(this.sequence_, undefined)
+    this.sequence_ = seq
```
Comment (faustbrian) [inline packages/core-transaction-pool-mem/lib/mem-pool-transaction.js]:
```suggestion
    this._sequence = seq
```

File: packages/core-transaction-pool-mem/lib/mem-pool-transaction.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
+'use strict'
+
+const assert = require('assert')
+const { Transaction } = require('@arkecosystem/crypto').models
+
+/**
+ * A mem pool transaction.
+ * A normal transaction + a sequence number used to order by insertion time.
+ */
+class MemPoolTransaction {
```
Comment (faustbrian) [inline packages/core-transaction-pool-mem/lib/mem-pool-transaction.js]:
```suggestion
module.exports = class MemPoolTransaction {
```

**Comment by vasild** [comment]
> Beware:
> 
> before this patch:
> 
> ```
>       ✓ multiple additions and retrievals (646ms)
>       ✓ delete + add after sync (173ms)
> ```
> 
> after this patch:
> 
> ```
>       ✓ multiple additions and retrievals (5779ms)
>       ✓ delete + add after sync (1306ms)
> ```

Just tried to sort on demand every time a sorted result is needed instead of storing the data sorted. We need two orders: by fee (to get highest fee transactions) and by expiration (to expire old transactions). This would make sense if we insert a lot (because then we do not do any sorting during insert) and read a little (because reading is expensive as it involves sorting):

```
      ✓ multiple additions and retrievals (29853ms)
      ✓ delete + add after sync (233ms)
```

Not good either.

File: packages/core-transaction-pool-mem/lib/mem-pool-transaction.js (hunk 7 above)
Comment (vasild) [reply packages/core-transaction-pool-mem/lib/mem-pool-transaction.js]: Done

File: packages/core-transaction-pool-mem/lib/mem-pool-transaction.js (hunk 6 above)
Comment (vasild) [reply packages/core-transaction-pool-mem/lib/mem-pool-transaction.js]: Done

File: packages/core-transaction-pool-mem/lib/mem-pool-transaction.js (hunk 1 above)
Comment (vasild) [reply packages/core-transaction-pool-mem/lib/mem-pool-transaction.js]: Done

File: packages/core-transaction-pool-mem/lib/mem-pool-transaction.js (hunk 2 above)
Comment (vasild) [reply packages/core-transaction-pool-mem/lib/mem-pool-transaction.js]: Done

File: packages/core-transaction-pool-mem/lib/mem-pool-transaction.js (hunk 3 above)
Comment (vasild) [reply packages/core-transaction-pool-mem/lib/mem-pool-transaction.js]: Done

File: packages/core-transaction-pool-mem/lib/mem-pool-transaction.js (hunk 5 above)
Comment (vasild) [reply packages/core-transaction-pool-mem/lib/mem-pool-transaction.js]: Done

**Comment by vasild** [comment]
> > Beware:
> > before this patch:
> > ```
> >       ✓ multiple additions and retrievals (646ms)
> >       ✓ delete + add after sync (173ms)
> > ```
> > after this patch:
> > ```
> >       ✓ multiple additions and retrievals (5779ms)
> >       ✓ delete + add after sync (1306ms)
> > ```
> 
> Just tried to sort on demand every time a sorted result is needed instead of storing the data sorted. We need two orders: by fee (to get highest fee transactions) and by expiration (to expire old transactions). This would make sense if we insert a lot (because then we do not do any sorting during insert) and read a little (because reading is expensive as it involves sorting):
> 
> ```
>       ✓ multiple additions and retrievals (29853ms)
>       ✓ delete + add after sync (233ms)
> ```
> 
> Not good either.

Using http://www.collectionsjs.com/sorted-set:
```
      ✓ multiple additions and retrievals (21493ms)
      ✓ delete + add after sync (177ms)
```

**Comment by vasild** [comment]
> @vasild what dataset size are you using that you end up with 21 seconds, millions of records or is the 1024 from the tests?

It is 1024 from the tests (unmodified), but this does not matter so much. What matters is the magnitude of the change: before patch <1s, with array, >5s. Simply not acceptable. Also, I would expect the sorted-map solution to be faster than the array, not 4x slower. I have not fully analyzed this, just sharing intermediate results.

**Comment by vasild** [comment]
Added one more test and changed the sorting to happen on demand / lazily.

Looks on par with the performance as of before order by fee (before this patch):

```
before:
      ✓ multiple additions and retrievals (635ms)
      ✓ delete + add after sync (173ms)
      ✓ add many then get first few (821ms)

after:
      ✓ multiple additions and retrievals (717ms)
      ✓ delete + add after sync (181ms)
      ✓ add many then get first few (865ms)
```

File: packages/core-transaction-pool-mem/package.json [hunk 8]
```diff
@@ -28,7 +28,8 @@
     "better-sqlite3": "^5.0.1",
     "collections": "^5.1.5",
     "delay": "^4.1.0",
-    "fs-extra": "^7.0.0"
+    "fs-extra": "^7.0.0",
+    "random-seed": "^0.3.0"
```
[the remark this answers is not in the document]
Comment (vasild) [reply packages/core-transaction-pool-mem/package.json]: Right, good catch! Moved. Thanks!

**Pull request merged**